okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -298,6 +299,8 @@ public void streamFailed(IOException e) {
           noNewStreams = true;
           route = null;
         }
+      } else if (e instanceof ConnectionShutdownException) {
+        noNewStreams = true;
```

Comment:
Or maybe even this:
```
noNewStreams = !(e instanceof StreamResetException)
    || ((StreamResetException) e).errorCode != ErrorCode.CANCEL;
```
ie. by default everything kills the connection _except_ for stream canceling, which we know is safe.